res/res_pjsip_refer.c

```diff
@@ -786,6 +791,498 @@ static void refer_blind_callback(struct ast_channel *chan, struct transfer_chann
+	pjsip_dialog *dlg;
+	struct refer_data *refer_data;
+	pjsip_uri *parsed_uri;
+	pjsip_sip_uri *refer_to_uri;
+
+	/*
+	 * If this is a request in response to a 401/407 Unauthorized challenge, the
+	 * Refer-To URI has been rewritten already, so don't attempt to re-write it again.
+	 * Checking for presence of the Authorization header is not an ideal solution. We do this because
+	 * there exists some race condition where this dialog is not the same as the one used
+	 * to send the original request in which case we don't have the correct refer_data.
+	 */
+	if (!refer_to_str
+		|| pjsip_msg_find_hdr(tdata->msg, PJSIP_H_AUTHORIZATION, NULL)
+		|| !(dlg = pjsip_tdata_get_dlg(tdata))
+		|| !(refer_data = pjsip_dlg_get_mod_data(dlg, refer_out_of_dialog_module.id))
+		|| !refer_data->to_self
+		|| !(parsed_uri = get_refer_to_uri(tdata))) {
+		goto out;
+	}
+	refer_to_uri = pjsip_uri_get_uri(parsed_uri);
+	ast_sip_rewrite_uri_to_local(refer_to_uri, tdata);
+
+	pjsip_uri_print(PJSIP_URI_IN_CONTACT_HDR, parsed_uri, ast_str_buffer(refer_to_str), ast_str_size(refer_to_str));
+	refer_to_hdr = pjsip_msg_find_hdr_by_name(tdata->msg, &REFER_TO, NULL);
+	pj_strdup2(tdata->pool, &refer_to_hdr->hvalue, ast_str_buffer(refer_to_str));
+
+out:
+	return PJ_SUCCESS;
+}
+
+static int refer_unreference_dialog(void *obj)
+{
+	struct refer_out_of_dialog *data = obj;
+
+	/* This is why we keep the dialog on the subscription. When the subscription
+	 * is destroyed, there is no guarantee that the underlying dialog is ready
+	 * to be destroyed. Furthermore, there's no guarantee in the opposite direction
+	 * either. The dialog could be destroyed before our subscription is. We fix
+	 * this problem by keeping a reference to the dialog until it is time to
+	 * destroy the subscription.
+	 */
+	pjsip_dlg_dec_session(data->dlg, &refer_out_of_dialog_module);
+	data->dlg = NULL;
+
+	return 0;
+}
+/*! \brief Destructor for REFER out of dialog structure */
+static void refer_out_of_dialog_destroy(void *obj) {
+	struct refer_out_of_dialog *data = obj;
+
+	if (data->dlg) {
+		ast_sip_push_task_wait_servant(refer_serializer, refer_unreference_dialog, data);
+	}
+}
+
+/*!
+ * \internal
+ * \brief Callback function to report status of implicit REFER-NOTIFY subscription.
+ *
+ * This function will be called on any state change in the REFER-NOTIFY subscription.
+ * Its primary purpose is to report SUCCESS/FAILURE of a refer initiated via
+ * \ref refer_send as well as to terminate the subscription, if necessary.
+ */
+static void refer_client_on_evsub_state(pjsip_evsub *sub, pjsip_event *event)
+{
+	pjsip_tx_data *tdata;
+	RAII_VAR(struct ast_sip_endpoint *, endpt, NULL, ao2_cleanup);
+	struct refer_out_of_dialog *refer_data;
+	int refer_success;
+	int res = 0;
+
+	if (!event) {
+		return;
+	}
+
+	refer_data = pjsip_evsub_get_mod_data(sub, refer_out_of_dialog_module.id);
+	if (!refer_data || !refer_data->dlg) {
+		return;
+	}
+
+	pjsip_dlg_inc_lock(refer_data->dlg);
```

Comment:
Why the lock here? What case is being protected?

Reply:
Now that you pointed it out, I don't think it is needed. And the critical section is huge... I'll remove it.